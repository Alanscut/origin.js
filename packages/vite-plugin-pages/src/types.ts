export type VitePluginVuePagesOptions = {
  excludes?: string[]
}
