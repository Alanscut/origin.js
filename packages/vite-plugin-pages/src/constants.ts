export const MODULE_NAME = 'pages'
