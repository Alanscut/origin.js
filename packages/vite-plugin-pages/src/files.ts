import * as globby from 'globby'
import { replaceWithSlash } from './utils'

function getExcludes(excludes: string[]) {
  return ['node_modules', '.git', '**/__*__/**', ...excludes]
}

export function getFiles(
  directory: string,
  extensions: string[],
  excludes?: string[],
): string[] {
  // setup excludes
  excludes = getExcludes((excludes || []).map(ex => replaceWithSlash(ex)))

  // return read files by extensions
  const pattern =
    extensions.length > 1
      ? `**/*.{${extensions.join(',')}}`
      : `**/*.${extensions[0]}`
  return globby.sync(pattern, {
    gitignore: true,
    ignore: excludes,
    cwd: replaceWithSlash(directory),
    onlyFiles: true,
  })
}

/**
 * sort files, make sure that index.vue is the first element of each directory
 * @param filePaths
 */
export function sortFilePaths(filePaths: string[]): string[] {
  filePaths.sort((a, b) => {
    a = a.endsWith('index.vue') ? a.slice(0, a.length - 9) : a
    b = b.endsWith('index.vue') ? b.slice(0, b.length - 9) : b
    if (a < b) {
      return -1
    }
    if (a > b) {
      return 1
    }
    return 0
  })
  return filePaths
}
