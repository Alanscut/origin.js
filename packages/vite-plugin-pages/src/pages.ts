import { join } from 'path'
import { Page } from './types'
import { getVueFiles } from './files'
import { replaceWithSlash } from './utils'

function getPage(pathFromPagesDir: string, pathFromRootDir: string): Page {
  return {
    pathFromPagesDir,
    pathFromRootDir,
  }
}

export function getPages(pagesDir: string): Page[] {
  const files = getVueFiles(pagesDir)
  return files.map(file =>
    getPage(file, replaceWithSlash(join(pagesDir, file))),
  )
}
