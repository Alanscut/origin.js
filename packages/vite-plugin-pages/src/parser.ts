import { vueSfcAstParser } from '@originjs/vue-sfc-ast-parser'
import fs from 'fs'
import { join, parse } from 'path'
import { getLayoutFiles } from './files'
import { PluginOptions, Route } from './types'

/**
 * get the __layout property from pages
 *
 * @param path file path in pages directory
 */
export function getLayoutProperties(path: string) {
  const source = fs.readFileSync(path, 'utf8')
  const { scriptAST, jscodeshiftParser } = vueSfcAstParser({ path, source })

  const layoutDeclare = scriptAST.find(jscodeshiftParser.VariableDeclarator, {
    id: {
      name: '__layout',
    },
  })
  let layout
  layoutDeclare.forEach(({ node }) => {
    // @ts-ignore
    layout = node?.init?.value
  })

  return layout
}

export function setLayout(routes: Route[], options: PluginOptions = {}) {
  const layouts = getLayoutMap(
    options.layoutsDir,
    options.extension,
    options.excludes,
  )

  routes.map(route => {
    // 当没有 __layout 属性，也没有 default 的时候，不要转换
    if (!route.meta?.layout && !hasDefaultLayout(options.layoutsDir)) {
      return route
    }
    return {
      path: route.path,
      component: layouts.get(route.meta?.layout) || 'default',
      children: [route],
    }
  })
}

function hasDefaultLayout(layoutDir: string): boolean {
  const layoutFiles = getLayoutFiles(layoutDir, ['vue'])
  for (const file of layoutFiles) {
    if (file === 'default.vue') {
      return true
    }
  }
  return false
}

function getLayoutMap(
  directory: string,
  extensions: string[],
  excludes?: string[],
) {
  const layoutFiles = getLayoutFiles(directory, extensions, excludes)
  const layoutMap = new Map()

  for (const file of layoutFiles) {
    const parsedFile = parse(file)
    const layoutPath = join(directory, file)
    layoutMap.set(parsedFile.name, `() => import('${layoutPath}')`)
  }

  return layoutMap
}
