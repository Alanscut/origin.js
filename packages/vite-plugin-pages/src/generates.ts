import { getLayoutProperties, setLayout } from './parser'
import { PluginOptions, Route, Page } from './types'

/**
 * Return whether the current route node name represents a dynamic route
 * @param routeNodeName: route node name
 */
function isDynamicNodeName(routeNodeName: string): boolean {
  return routeNodeName.startsWith('_')
}

/**
 * Return whether the current route node name represents a catch all route
 * @param routeNodeName: route node name
 */
function isCatchAllNodeName(routeNodeName: string): boolean {
  return routeNodeName === '_'
}

function prepareRoutes(routes: Route[], parentRoute?: Route): Route[] {
  for (const route of routes) {
    if (parentRoute && route.path.startsWith('/')) {
      route.path = route.path.slice(1)
    }
    if (route.children && route.children.length > 0) {
      route.children = prepareRoutes(route.children, route)
    }
  }
  return routes
}

export function generateRoutes(
  pages: Page[],
  options: PluginOptions = {},
): Route[] {
  let routes: Route[] = []

  pages.forEach(page => {
    const { pathFromPagesDir } = page
    // remove file extension, split with '/' to get route nodes
    const routeNodeNames = pathFromPagesDir.split('.')[0].split('/')

    const route: Route = {
      name: '',
      path: '',
      component: `${page.pathFromRootDir}`,
      meta: { layout: getLayoutProperties(pathFromPagesDir) },
    }

    let parentRoutes = routes
    for (const routeNodeName of routeNodeNames) {
      const isDynamic = isDynamicNodeName(routeNodeName)
      const isCatchAll = isCatchAllNodeName(routeNodeName)

      route.name += route.name ? `-${routeNodeName}` : routeNodeName
      const parentRoute = parentRoutes.find(pRoute => pRoute.name == route.name)

      if (parentRoute) {
        // init parentRoute.children, update parentRoutes, continue to next loop
        parentRoute.children = parentRoute.children || []
        parentRoutes = parentRoute.children
        route.path = ''
        continue
      }

      if (isDynamic) {
        if (isCatchAll) {
          route.path = '/:pathMatch(.*)*'
        } else {
          route.path += `/:${routeNodeName.slice(1)}`
        }
      } else {
        route.path += `/${routeNodeName}`
      }
    }
    parentRoutes.push(route)
  })

  routes = prepareRoutes(routes)
  // find and only keep first catch all route, move it to last
  const catchAllRoute = routes.find(route => route.path === '/:pathMatch(.*)*')
  if (catchAllRoute) {
    routes = routes.filter(route => route.path !== '/:pathMatch(.*)*')
    routes.push(catchAllRoute)
  }

  // set layouts to routes
  setLayout(routes, options)

  return routes
}
