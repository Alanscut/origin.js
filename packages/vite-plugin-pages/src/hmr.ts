import { ViteDevServer } from 'vite'
import { MODULE_NAME } from './constants'
import { updatePages } from './pages'
import { PluginOptions } from './types'
import { join } from 'path'
import { updateLayouts } from './parser'
import { replaceWithSlash } from './utils'

function getModule(server: ViteDevServer) {
  const { moduleGraph } = server
  const module = moduleGraph.getModuleById(MODULE_NAME)
  if (module) {
    moduleGraph.invalidateModule(module)
    return module
  }
  return
}

export function handleHmr(server: ViteDevServer, options: PluginOptions) {
  const { ws, watcher } = server
  const { root, pagesDir, layoutsDir } = options
  const pagesDirFullPath = replaceWithSlash(join(root, pagesDir)) + '/'
  const layoutsDirFullPath = replaceWithSlash(join(root, layoutsDir)) + '/'

  function reloadRoutes() {
    getModule(server)
    ws.send({
      type: 'full-reload',
    })
    return
  }

  function isPageFile(filePath: string): boolean {
    return filePath.startsWith(pagesDirFullPath)
  }

  function isLayoutFile(filePath: string): boolean {
    return filePath.startsWith(layoutsDirFullPath)
  }

  function getPathFromPagesDir(filePath: string): string {
    return filePath.replace(pagesDirFullPath, '')
  }

  function getPathFromLayoutsDir(filePath: string): string {
    return filePath.replace(layoutsDirFullPath, '')
  }

  const events = ['add', 'unlink', 'change']
  events.forEach(event => {
    watcher.on(event, async file => {
      file = replaceWithSlash(file)
      if (isPageFile(file)) {
        console.log('pages', event, file)
        const pathFromPagesDir = getPathFromPagesDir(file)
        updatePages(event, pagesDir, pathFromPagesDir)
        reloadRoutes()
      } else if (isLayoutFile(file)) {
        console.log('layouts', event, file)
        const pathFromLayoutsDir = getPathFromLayoutsDir(file)
        updateLayouts(event, layoutsDir, pathFromLayoutsDir)
        reloadRoutes()
      }
    })
  })
}
