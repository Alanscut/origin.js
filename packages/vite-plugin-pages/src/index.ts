import { Page, PluginOptions } from './types'
import { Plugin } from 'vite'
import { MODULE_NAME } from './constants'
import { generateRoutes } from './generates'
import { getPages } from './pages'

export default (userOptions: PluginOptions = {}): Plugin => {
  const DEFAULT_OPTIONS: PluginOptions = {
    root: process.cwd(),
    pagesDir: 'src/pages',
    extension: ['vue'],
  }
  const options: PluginOptions = Object.assign({}, DEFAULT_OPTIONS, userOptions)
  let pages: Page[]

  return {
    name: 'vite:pages',
    enforce: 'pre',
    configResolved(config) {
      options.root = config.root
      const pagesDir = 'src/pages'
      pages = getPages(pagesDir)
    },
    resolveId(id) {
      if (id === MODULE_NAME) {
        return id
      }
      return null
    },
    async load(id) {
      if (id !== MODULE_NAME) {
        return
      }
      const routes = generateRoutes(pages)
      return { routes }
    },
  }
}

export { PluginOptions }
