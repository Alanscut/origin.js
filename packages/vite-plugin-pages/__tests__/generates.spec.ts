import { generateRoutes } from '../src/generates'
import { Page, PluginOptions } from '../src/types'

describe('generates', () => {
  test('generateRoutes', () => {
    const pages: Page[] = [
      {
        pathFromPagesDir: '_.vue',
        pathFromRootDir:
          'packages/vite-plugin-pages/__tests__/assets/pages/_.vue',
      },
      {
        pathFromPagesDir: 'index.vue',
        pathFromRootDir:
          'packages/vite-plugin-pages/__tests__/assets/pages/index.vue',
      },
      {
        pathFromPagesDir: 'user/index.vue',
        pathFromRootDir:
          'packages/vite-plugin-pages/__tests__/assets/pages/user/index.vue',
      },
      {
        pathFromPagesDir: 'user/foo.vue',
        pathFromRootDir:
          'packages/vite-plugin-pages/__tests__/assets/pages/user/foo.vue',
      },
      {
        pathFromPagesDir: 'user/_id.vue',
        pathFromRootDir:
          'packages/vite-plugin-pages/__tests__/assets/pages/user/_id.vue',
      },
      {
        pathFromPagesDir: 'user/nest_user/foo.vue',
        pathFromRootDir:
          'packages/vite-plugin-pages/__tests__/assets/pages/user/nest_user/foo.vue',
      },
    ]
    const options: PluginOptions = {
      root: process.cwd(),
      pagesDir: 'packages/vite-plugin-pages/__tests__/assets/pages',
      layoutsDir: 'packages/vite-plugin-pages/__tests__/assets/layouts',
      extensions: ['vue'],
    }

    expect(generateRoutes(pages, options)).toMatchSnapshot()
  })
})
