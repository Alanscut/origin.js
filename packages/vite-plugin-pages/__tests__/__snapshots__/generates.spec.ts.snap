// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates generateRoutes 1`] = `
Array [
  Object {
    "component": "/index.vue",
    "name": "index",
    "path": "/",
  },
  Object {
    "component": "/user/index.vue",
    "name": "user-index",
    "path": "/user",
  },
  Object {
    "component": "/user/foo.vue",
    "name": "user-foo",
    "path": "/user/foo",
  },
  Object {
    "children": Array [
      Object {
        "component": "/user/_id/profile.vue",
        "name": "user-_id-profile",
        "path": "profile",
      },
    ],
    "component": "/user/_id.vue",
    "name": "user-_id",
    "path": "/user/:id",
  },
  Object {
    "component": "/user/nest_user/foo.vue",
    "name": "user-nest_user-foo",
    "path": "/user/nest_user/foo",
  },
  Object {
    "component": "/_.vue",
    "name": "_",
    "path": "/:pathMatch(.*)*",
  },
]
`;
